docs/pywbemcligeneraloptions.rst
.. _`Using the pywbemcli command line general options`:

Using the pywbemcli command line general options
================================================


.. _`Oveview of the general options`:

Overview of the general options
-------------------------------

General options define:

* The connection to the WBEM server against which the pywbemcli commands are to be
  executed (i.e url, default CIM namespace, security parameters, etc.)
  or the alternative mock of a WBEM server.
  See :ref:`Defining the WBEM server` for details.
* Operation behavior of pywbemcli for requests; i.e. using the pull operations
  vs the :term:`traditional operations`, or the client side timeout.
  See :ref:`Controlling operation behavior and monitoring operations` for
  details.
* Execution options to monitor the requests and responses (statistics keeping,
  and logging of operations) and to display the results.
  See :ref:`Controlling operation behavior and monitoring operations` and
  :ref:`Controlling output formats` for details.
* General options to show the pywbemcli version, and display help for any of
  the command groups or commands.

General options must be entered on the command line before the command group
or command.

For example the following enumerates qualifier declarations and outputs the
result in the ``simple`` table format:

.. code-block:: text

    pywbemcli --output-format simple --server http://localhost qualifier enumerate
    Qualifier Declarations
    Name         Type     Value    Array    Scopes       Flavors
    -----------  -------  -------  -------  -----------  ---------------
    Description  string            False    ANY          EnableOverride
                                                         ToSubclass
                                                         Translatable
    Key          boolean  False    False    PROPERTY     DisableOverride
                                            REFERENCE    ToSubclass

In interactive mode, general options from the command line are defined
once and retain their value throughout the execution of the interactive mode.

General options defined in the command line may be overwritten by specifying
them on interactive mode commands. The resulting new value is used only for
that one command, and is not retained between interactive mode command
executions.

In the following example, the :ref:`--output-format general option` is used
on an interactive mode command to overwrite the default output format, and the
:ref:`--server general option` is used to overwrite the server that was
specified in the command line invocation:

.. code-block:: text

    $ pywbemcli --server http://myserver

    pywbemcli> --output-format table --server http://localhost qualifier enumerate
    Qualifier Declarations
    +-------------+---------+---------+---------+-----------+-----------------+
    | Name        | Type    | Value   | Array   | Scopes    | Flavors         |
    +=============+=========+=========+=========+===========+=================+
    | Description | string  |         | False   | ANY       | EnableOverride  |
    |             |         |         |         |           | ToSubclass      |
    |             |         |         |         |           | Translatable    |
    +-------------+---------+---------+---------+-----------+-----------------+
    | Key         | boolean | False   | False   | PROPERTY  | DisableOverride |
    |             |         |         |         | REFERENCE | ToSubclass      |
    +-------------+---------+---------+---------+-----------+-----------------+


.. _`Defining the WBEM server`:

Defining the WBEM server
^^^^^^^^^^^^^^^^^^^^^^^^

The target WBEM server can be defined on the command line in several ways with
the following arguments :

1. Define connection information for a WBEM server by using the
   :ref:`--server general option` that specifies the URL of the server.

   The following general options can be used for specifying additional
   information for the connection:

   * The :ref:`--default-namespace general option` defines the default namespace
     to be used if a command does not specify a namespace.
   * The :ref:`--user general option` defines the user name for authenticating
     with the WBEM server.
   * The :ref:`--password general option` defines the password for
     authenticating with the WBEM server.
   * The :ref:`--verify general option` defines whether the client verifies
     certificates received from the WBEM server.
   * The :ref:`--certfile general option` defines the client certificate file.
   * The :ref:`--keyfile general option` defines the client key file.
   * The :ref:`--ca-certs general option` defines a collection of certificates
     against which certificates received from the WBEM server are verified.
   * The :ref:`--timeout general option` defines the client side timeout
     for operations.

2. Define a mock WBEM server by using the :ref:`--mock-server general option`.

   The mock WBEM server is part of pywbemcli and allows testing or
   demonstrating pywbemcli without having access to a real WBEM server.
   For details, see :ref:`Mock WBEM server support`.

3. Refer to a persisted connection definition for either a WBEM server or
   mock WBEM server by using the :ref:`--name general option` that
   specifies the name of the connection definition.

   Persisted connection definitions are stored in a :term:`connections file`
   and are managed with the :ref:`Connection command group`.


.. _`Controlling operation behavior and monitoring operations`:

Controlling operation behavior and monitoring operations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Several general options control the behavior and monitoring of the operations
executed against the WBEM server.

The DMTF specifications allow alternative forms of some operations,
pywbemcli implements this flexibility and controls the choice of either the
pull operations or the traditional operations through the :ref:`--use-pull
general option`. With this option the user can choose to use either type of
operation if that operation is available on the WBEM server. See
:ref:`Pywbemcli and the DMTF pull operations` for more information on pull
operations.

Since the pull operations include the ability to select the maximum size of
returned chunks of data, the :ref:`--pull-max-cnt general option` can be used
to control response chunk sizes.

In many cases it is important to the user to be able to monitor details of the
operations executed against the WBEM server, either the APIs executed in pywbem,
or the HTTP requests and responses and the time statistics for these
operations.

The :ref:`--log general option` provides the capability to log information about
this flow including:

* The API calls and responses/exceptions executed by pywbem.
* the HTTP requests and responses that pass between pywbemcli and the WBEM server.

The :ref:`--log general option` configures the logging including what is logged
and the destination for the log output.

Thus, for example, the following command enumerates qualifiers and writes
the log entries for the CIM-XML HTTP requests and responses to the pywbemcli log
file ``pywbemcli.log``:

.. code-block:: text

   $ pywbemcli --output-format table --server http://localhost --log http=file qualifier enumerate
   <displays table of qualifier declarations>

   $ cat pywbemcli.log
   2019-09-16 21:14:04,296-pywbem.http.1-21020-Connection:1-21020 WBEMConnection(url='http://localhost', ...)
   2019-09-16 21:14:04,297-pywbem.http.1-21020-Request:1-21020 POST /cimom 11 http://localhost
       CIMOperation:'MethodCall' CIMMethod:'EnumerateQualifiers' CIMObject:u'root/cimv2'
       <?xml version="1.0" encoding="utf-8" ?>
       <CIM CIMVERSION="2.0" DTDVERSION="2.0"><MESSAGE ID="1001" PROTOCOLVERSION="1.0">
       ...
       </CIM>
   . . .


.. _`Controlling output formats`:

Controlling output formats
^^^^^^^^^^^^^^^^^^^^^^^^^^

Pywbemcli supports multiple output formats for command results by using the
:ref:`--output-format general option`.

The output formats fall into three groups (table formats, CIM object formats,
and a tree format); however, not all formats are supported or applicable for all
commands. For more details, see :ref:`Output formats`.


.. _`Other miscellaneous general options`:

Other miscellaneous general options
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :ref:`--verbose general option` displays extra information about the pywbemcli
internal processing, the :ref:`--version general option` displays pywbemcli version
information and the :ref:`--help general option` provides top level help


.. _`General options descriptions`:

General options descriptions
----------------------------

This section defines in detail the requirements, characteristics, and any special
syntax of each general option.


.. _`--server general option`:

--server general option
^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--server``/``-s`` general option is a string that is
the URL of the WBEM server to which pywbemcli will connect, in the format:

.. code-block:: text

    [SCHEME://]HOST[:PORT]

Where:

* **SCHEME**: The protocol to be used. Must be "https" or "http". Default: "https".
* **HOST**: The WBEM server host. Must be a short hostname, fully qualified DNS
  hostname, literal IPv4 address, or literal IPv6 address.
  See :term:`RFC3986` and :term:`RFC6874` for details.
* **PORT**: The WBEM server port to be used.
  Default: 5988 for HTTP, and 5989 for HTTPS.

This option is mutually exclusive with the :ref:`--name general option` and the
:ref:`--mock-server general option` since each defines a connection to a WBEM
server.

In the interactive mode the connection is not actually established until a
command requiring access to the WBEM server is entered.

Examples for the argument value of this option include:

.. code-block:: text

    https://localhost:15345       # https, port 15345, hostname localhost
    http://10.2.3.9               # http, port 5988, IPv4 address 10.2.3.9
    https://10.2.3.9              # https, port 5989, IPv4 address 10.2.3.9
    http://[2001:db8::1234-eth0]  # http, port 5988, IPv6 address 2001:db8::1234, interface eth0


.. _`--name general option`:

--name general option
^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--name``/``-n`` general option is a string that is
the name of a connection definition in the :term:`connections file`.
The parameters for this named connection definition will be loaded from the
:term:`connections file` to become the current WBEM connection in pywbemcli.

In the interactive mode the connection is not actually established until a
command requiring access to the WBEM server is entered.

This option is mutually exclusive with :ref:`--server general option` and
:ref:`--mock-server general option` since each defines a connection to a WBEM
server.

The following example creates a connection definition named ``myserver``
in the connections file, and then uses that connection to execute
``class get``:

.. code-block:: text

    $ pywbemcli --server http://localhost --user me --password mypw connection save myserver

    $ pywbemcli --name myserver class get CIM_ManagedElement
    <displays MOF for CIM_ManagedElement>

See :ref:`Connection command group` for more information on managing
connections.


.. _`--default-namespace general option`:

--default-namespace general option
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--default-namespace``/``-d`` general option is a
string that defines the default :term:`CIM namespace` to use for the target
WBEM server.

If this option is not specified, the default namespace will be ``root/cimv2``.

The default namespace will be used if the ``--namespace``/``-n`` command
option is not used on a command.

Some commands execute against multiple or all namespaces, for example the
the ``class find`` command.


.. _`--user general option`:

--user general option
^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--user``/``-u`` general option is a string that is
the user name for authenticating with the WBEM server.


.. _`--password general option`:

--password general option
^^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--password``/``-p`` general option is a string that is the
password for authenticating with the WBEM server.

This option is normally required if the
:ref:`--user general option` is defined.  If passwords are saved into the
:term:`connections file` they are not encrypted in the file.

If the WBEM operations performed by any pywbemcli command require a password,
the password is prompted for if ``--user``/``-u`` is used (in both modes of
operation) and ``--password``/``-p`` is not used.

.. code-block:: text

    $ pywbemcli --server http://localhost --user me class get
    Enter password: <password>
    . . . <The display output from get class>

If both the ``--user``/``-u`` and ``--password``/``-p`` general options are
used, the command is executed without a password prompt:

.. code-block:: text

    $ pywbemcli --server http://localhost --user me --password blah class get
    . . . <The display output from get class>

If the operations performed by a particular pywbemcli command do not
require a password or no user is supplied, no password is prompted.
For example:

.. code-block:: text

      $ pywbemcli --help
      . . . <help output>

For script integration, it is important to have a way to avoid the interactive
password prompt. This can be done by storing the password string in an
environment variable or specifying it on the command line.
See :ref:`Environment variables for general options`.

The pywbemcli :ref:`Connection export command` outputs the (bash/Windows)
shell commands to set all needed environment variables.

The environment variable output is OS dependent. Thus for example in Unix type
OSs:

.. code-block:: text

    $ pywbemcli --server http://localhost --user fred connection export
    export PYWBEMCLI_SERVER=http://localhost
    export PYWBEMCLI_NAMESPACE=root/cimv2
    ...

This ability can be used to set those environment variables and thus to persist
the connection name in the shell environment, from where it will be used in
any subsequent pywbemcli commands:

.. code-block:: text

    $ eval $(pywbemcli --server http://localhost --user fred connection export)

    $ env | grep PYWBEMCLI
    export PYWBEMCLI_SERVER=http://localhost
    export PYWBEMCLI_NAMESPACE=root/cimv2

    $ pywbemcli server namespaces
    . . . <list of namespaces for the defined server>


.. _`--timeout general option`:

--timeout general option
^^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--timeout``/``-t`` general option is an integer that
defines the
client side timeout in seconds. The pywbem client includes a timeout mechanism
that closes a WBEM connection if there is no response to a request to the WBEM
server in the time defined by this value. Pywbemcli defaults to a
predefined timeout (normally 30 seconds) if this option is not defined.


.. _`--verify general option`:

--verify general option
^^^^^^^^^^^^^^^^^^^^^^^

The pair of ``--verify`` and ``--no-verify`` general options control whether or
not the client verifies any certificates received from the WBEM server.

By default or if ``--verify`` is specified, any certificates returned by the
server are verified. If ``--no-verify`` is specified, any certificates returned
by the server are accepted without verification.

This general option uses the approach with two long option names to allow the
user to specifically enable or disable certificate verification when this
general option is used in interactive mode.


.. _`--certfile general option`:

--certfile general option
^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--certfile`` general option is the file path of a
PEM file containing a X.509 client certificate to be presented to the WBEM
server during the TLS/SSL handshake, enabling 2-way (mutual) authentication.
This option is used only with HTTPS.

If ``--certfile`` is not used, no client certificate is presented to the server,
resulting in 1-way authentication during the TLS/SSL handshake.

For more information on authentication types, see:
https://pywbem.readthedocs.io/en/stable/client/security.html#authentication-types


.. _`--keyfile general option`:

--keyfile general option
^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--keyfile`` general option is the file path of a
PEM file containing the private key belonging to the public key that is
part of the X.509 certificate. See :ref:`--certfile general option` for more
information.

Not required if the private key is part of the file defined in the
:ref:`--certfile general option`. ``keyfile`` is not allowed if
:ref:`--certfile general option` is not provided. Default: No client
key file. The client private key should then be part of the file defined by
``--certfile``.


.. _`--ca-certs general option`:

--ca-certs general option
^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--ca-certs`` general option is the path name of a file or
directory containing certificates that will be matched against a certificate
received from the WBEM server. The default is OS dependent and is a set of
system directories where certificates are expected to be stored for the client
OS.

Specifying the ``--no-verify`` general option (see :ref:`--verify general option`)
bypasses client verification of the WBEM server certificate.


.. _`--timestats general option`:

--timestats general option
^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``--timestats`` general option is a boolean option that enables the
gathering and display of time
statistics on the interactions with the WBEM server.  If enabled, the
time statistics are output after each command is executed including the
operations executed, the size of the operations, and the execution time.


.. _`--use-pull general option`:

--use-pull general option
^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--use-pull``/``-u`` general option determines
whether the pull operations or :term:`traditional operations` are used for the
``instance enumerate``, ``instance references``, ``instance associators``
and ``instance query`` commands. See :ref:`Pywbemcli and
the DMTF pull operations` for more information on pull operations. The choices
for the argument value are as follows:

* ``yes`` - pull operations will be used and if the server does not
  support pull, the request will fail.
* ``no`` - forces pywbemcli to try only the traditional non-pull operations.
* ``either`` - (default) pywbem tries both; first pull operations and then
  :term:`traditional operations`.


.. _`--pull-max-cnt general option`:

--pull-max-cnt general option
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--pull-max-cnt`` general option is an integer
passed to the WBEM server with the open and pull operation requests.
This integer tells the server the maximum number of objects
to be returned for each pull request if pull operations are used. This must
be a positive non-zero integer. The default is 1000. See :ref:`Pywbemcli and the
DMTF pull operations` for more information on pull operations.


.. _`--mock-server general option`:

--mock-server general option
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--mock-server``/``-m`` general option is a file
path of a MOF file or Python script that loads a mock WBEM server in the
pywbemcli process with mock data (i.e. CIM objects).
This allows pywbemcli to be used without access to a real WBEM server.

This option may be specified multiple times.

This option may be specified multiple times to define multiple MOF and Python
files that make up the definition of a mock server. The files must have the
suffix ".mof" for MOF files and ".py" for Python scripts.

When this option is used, the security options (ex. ``--user``) are irrelevant;
they may be specified but are not used.

The following example creates a mock server with two files defining the mock
data, shows what parameters are defined for the connection, and then saves that
connection named ``mymockserver``:

.. code-block:: text

    $ pywbemcli --mock-server classdefs.mof --mock-server insts.py --default-namespace root/myhome

    pywbemcli> connection show
    name: not-saved (current)
      server:
      mock-server: classdefs.mof, insts.py
      . . .

    pywbemcli> connection save mymockserver

See chapter :ref:`Mock WBEM server support` for more information on defining
the files for a mock server.


.. _`--output-format general option`:

--output-format general option
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The argument value of the ``--output-format``/``-o`` general option is a string
that defines the output format in which the result of any pywbemcli commands
is displayed. The default output format depends on the command.

For details, see :ref:`Output formats`.


.. _`--log general option`:

--log general option
^^^^^^^^^^^^^^^^^^^^

The argument value of the  ``--log``/``-l`` general option defines the destination and
parameters of logging of the requests and responses to the WBEM server.

For details, see :ref:`Pywbemcli defined logging`.


.. _`--verbose general option`:

--verbose general option
^^^^^^^^^^^^^^^^^^^^^^^^

The ``--verbose``/``-v`` general option is a boolean option that enables the
display of extra information about the processing.


.. _`--pdb general option`:

--pdb general option
""""""""""""""""""""

The ``--pdb`` general option is a boolean option that enables debugging
with the built-in pdb debugger.

If debugging is enabled, execution of the pywbemcli command will pause just
before the command within pywbemcli is executed, and the pdb debugger prompt
will appear. See `pdb debugger commands`_ for details on how to operate the
built-in pdb debugger.

.. _`pdb debugger commands`: https://docs.python.org/2.7/library/pdb.html#debugger-commands


.. _`--version general option`:

--version general option
^^^^^^^^^^^^^^^^^^^^^^^^

The ``--version``/``-V`` general option displays the version of this command
and of the pywbem package it uses, and then exits.


.. _`--help general option`:

--help general option
^^^^^^^^^^^^^^^^^^^^^

The ``--help``/``-h`` general option displays help text which describes the
command groups and general options, and then exits.


.. _`Environment variables for general options`:

Environment variables for general options
-----------------------------------------

Pywbemcli defines environment variables corresponding to its general options
as follows:

==============================  ============================
Environment variable            Corresponding general option
==============================  ============================
PYWBEMCLI_SERVER                ``--server``
PYWBEMCLI_NAME                  ``--name``
PYWBEMCLI_USER                  ``--user``
PYWBEMCLI_PASSWORD              ``--password``
PYWBEMCLI_OUTPUT_FORMAT         ``--output-format``
PYWBEMCLI_DEFAULT_NAMESPACE     ``--default-namespace``
PYWBEMCLI_TIMEOUT               ``--timeout``
PYWBEMCLI_KEYFILE               ``--keyfile``
PYWBEMCLI_CERTFILE              ``--certfile``
PYWBEMCLI_CACERTS               ``--ca-certs``
PYWBEMCLI_USE_PULL              ``--use-pull``
PYWBEMCLI_PULL_MAX_CNT          ``--pull-max-cnt``
PYWBEMCLI_STATS_ENABLED         ``--timestats``
PYWBEMCLI_MOCK_SERVER (1)       ``--mock-server``
PYWBEMCLI_LOG                   ``--log``
PYWBEMCLI_PDB                   ``--pdb``
==============================  ============================

Notes:

(1) The ``--mock-server`` general option can be specified multiple times. To
    do that with the PYWBEMCLI_MOCK_SERVER environment variable, separate
    the multiple path names with space.

If these environment variables are set, the corresponding general options
default to the value of the environment variables.
If both an environment variable and its corresponding general option are
set, the command line option overrides the environment variable with no
warning.
Environment variables are not provided for command options or command arguments.

In the following example, the pywbemcli command uses server
``http://localhost`` defined by the environment variable:

.. code-block:: text

      $ export PYWBEMCLI_SERVER=http://localhost
      $ pywbemcli class get CIM_ManagedElement
        <displays MOF for CIM_ManagedElement>

The pywbemcli :ref:`Connection export command` outputs the (bash/Windows)
shell commands to set all needed environment variables:

.. code-block:: text

    $ pywbemcli --server http://localhost --user fred connection export
    export PYWBEMCLI_SERVER=http://localhost
    export PYWBEMCLI_NAMESPACE=root/cimv2
    . . .

This ability can be used to set those environment variables and thus to persist
the connection name in the shell environment, from where it will be used in
any subsequent pywbemcli commands:

.. code-block:: text

    $ eval $(pywbemcli --server http://localhost --user fred)

    $ env | grep PYWBEMCLI
    export PYWBEMCLI_SERVER=http://localhost
    export PYWBEMCLI_NAMESPACE=root/cimv2
    . . .

    $ pywbemcli server namespaces
    . . . <list of namespaces for the defined server>


.. _`Pywbemcli and the DMTF pull operations`:

Pywbemcli and the DMTF pull operations
--------------------------------------

The DMTF specifications and pywbem includes two ways to execute the enumerate
instance type operations (``Associators``, ``References``,``
EnumerateInstances``, ``ExecQuery``):

* The :term:`traditional operations` (ex. ``EnumerateInstances``)

docs/changes.rst

.. _`Change log`:

Change log
==========


pywbemtools 0.6.0
-----------------

This version is currently in development and is shown as |version|.

This version contains all fixes up to pywbemtools 0.5.0.

Released: not yet

**Incompatible changes:**

**Deprecations:**

**Bug fixes:**

* Fix issue with mixed old and new formats on click.echo statement.
  (See issue #419)

* Fixed missing Python 3.7 in supported environments shown on Pypi.
  (See issue #416)

* Fixed that the 'class find' command showed the --namespace option twice
  (see issue #417)

* Added PyYAML>=5.1 as a prerequisite package for pywbemtools for installation.
  So far, it was pulled in indirectly via pywbem.

* Fixed case sensitive matching of class names in instance modify by
  picking up the fix in pywbem 0.14.6. (See issue #429).

* Fixed issue where extra diagnostic information about click was being displayed
  when the general option --verbose was defined.

* Fixed issue with x509 parameter of WBEMConnection. (See issue #468)

* Fixed issue with class find command not returning connection error when
  cannot connect to server.

* Added documentation for the --version general option.

* Increased pywbem minimum version to 0.16.0 to accomodate install issues
  on Python 3.4, and to pick up other fixes.

* Test: Accomodated new formatting of error messages in Click 7.1.1, and
  excluded Click 7.1 due to bug.

* Test: Fixed dependency to Python development packages on CygWin platform
  in Appveyor CI.

* Pygments 2.4.0 and readme-renderer 25.0 have removed support for Python 3.4
  and have therefore been pinned to below these versions on Python 3.4.

* Fix bug where order of commands listed in help output was different for
  different versions of Python. (See issue # 510)

* Increased minimum version of pluggy package from 0.12.0 to 0.13.0
  because it failed during loading of pytest plugins on Python 3.8.
  (See issue #494)

* Test: Changed testcases that check the CIM-XML generated with output format
  'xml' to tolerate the different order of XML attributes that happens on
  Python 3.8 (See issue #494)

* Fixed several badges on the README page.

**Enhancements:**

* Promoted development status of pywbemtools from Alpha to Beta.
  (See issue #476)

* Add capability to reorder commands in the help for each group.  The commands
  in all groups except for the top group (pywbemcli -h) are ordered in the
  help list by their order in their source file. The display of commands in
  the top level group is alphabetical except that connection, help, and repl
  are reordered to the bottom of the list. (See issue #466)

* Define alternatives for creating INSTANCENAME input parameter since the
  original form using, WBEMURI is error prone with quote marks.
  (see issue #390)

* Add prompt-toolkit auto-suggest.  This extends the command completion
  capability in the repl mode (interactive mode) to make suggestions on
  command line input based on the history file.  Usually auto-suggest completion
  will be shown as gray text behind the current input. Auto-suggest is not
  available in command line mode.

* Add ability to filter results of 'class enumerate', 'class find', and
  'instance count' commands for selected class qualifiers.  This ability
  is based on 3 new options for each of the above commands '--association',
  '--indication', and '--experimental' each of which has a corresponding
  'no-...'. The user can filter to find classes with combinations of these
  options returning only classes that meet the option criteria.  Thus,
  '--association' returns classes that are associations and '--no-association'
  returns only classes that are not associations.  See issue # 447

* Added support for disabling the spinner that is displayed by default during
  any ongoing activities, by setting the env var PYWBEM_SPINNER to 'false',
  '0', or the empty string. This is useful when debugging.  See issue #465.

* Modified the response handling on commands that may return nothing with
  successful response to display a message if the general option --verbose
  is defined and display nothing if --verbose not set.  This includes
  class/instance delete, instance modify and the commands that display
  cim objects. (See issue #123)

* Changed the `--ca-certs` general option to support the changes as of
  pywbem version 1.0.0 (new values 'system' and 'certifi', and default changed
  from a fixed set of directories to 'certifi'). The pywbem version is
  determined at run time and pywbem versions before 1.0.0 are still supported.

* Test: Improved assertion messages in tests.

* Output of "pywbemcli server profiles" command is now reliably sorted by
  version in addition to org and name. (See issue #500)

* Added support for a new `--pdb` general option and corresponding
  `PYWBEMCLI_PDB` environment variable that causes the pywbemcli command
  to come up with the pdb debugger before invoking the specified command.
  This is a debug feature that is expected to be used mainly by the
  developers of pywbemcli. (See issue #505)

* Test: Added support for entering the pdb debugger from specific unit testcases
  by setting the condition parameter of the testcase to the string 'pdb'.
  This causes pywbemcli to be invoked with the new --pdb option for that
  testcase. (See issue #505)

* Removed usage of the "pbr" package. It was used for automatic versioning
  of the pywbemtools package, but it had too many drawbacks for that small
  benefit (See issue #178):

  - Getting the minor version increased in a development version (instead of
    the patch version) by means of markers in the commit message never worked.
  - The package needed to be tagged twice during the release process.
  - If the last tag was too far in the past of the commit history, the
    shallow git checkout used by Travis failed and its depth needed to be
    adjusted. At some point this defeats the purpose of a shallow checkout.


pywbemtools/pywbemcli/config.py
# (C) Copyright 2017 IBM Corp.
# (C) Copyright 2017 Inova Development Inc.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Pywbemcli supports a very limited number of configuration variables that
influence certain specific behavior.

These configuration variables are read by pywbem only after its modules have
been loaded, so they can be modified by the user directly after importing
pywbem. For example:

::

    import pywbemcli
    pywbem.DEFAULT_CONNECTION_TIMEOUT = 40

Note that the pywbem source file defining these variables should not be changed
by the user. Instead, the technique shown in the example above should be used to
modify the configuration variables.

Note: Due to limitations of the documentatin tooling, the following
configuration variables are shown in the ``pywbemcli.config`` namespace.
However, they should be used from the ``pywbemcli`` namespace.
"""

# This module is meant to be safe for 'import *'.


__all__ = ['DEFAULT_CONNECTION_TIMEOUT', 'DEFAULT_OUTPUT_FORMAT',
           'DEFAULT_NAMESPACE', 'PYWBEMCLI_PROMPT', 'PYWBEMCLI_HISTORY_FILE',
           'DEFAULT_MAXPULLCNT', 'MAX_TIMEOUT', 'DEFAULT_URL_SCHEME',
           'USE_TERMINAL_WIDTH', 'DEFAULT_TABLE_WIDTH']

#: Default value in seconds for a WBEMConnection to timeout if the value
#: is not set by an input parameter.
#: Positive integer.
DEFAULT_CONNECTION_TIMEOUT = 30

#: Specifies the default output format selected if no output format is
#: defined on the cmd line, environment variable, or a config file.
DEFAULT_OUTPUT_FORMAT = 'simple'

#: Specifies the default namespace uses if no default namespace is defined
#: on the cmd line, environment variable, or a config file.
DEFAULT_NAMESPACE = 'root/cimv2'

#: Specifies the default query language to be used for exedquery operations
#: when a query language is not specified in the request or config
DEFAULT_QUERY_LANGUAGE = 'DMTF:CQL'

#: Characters for cmdline prompt when the pywbemcli repl is executing.
#: The prompt is presented at the beginning of a line awaiting a command
#: input.
#: The prompt MUST BE Unicode (prompt-toolkit requirement)
PYWBEMCLI_PROMPT = u'pywbemcli> '

#: File path of history file for interactive mode.
#: If the file name starts with tilde (which is handled by the shell, not by
#: the file system), it is properly expanded.
PYWBEMCLI_HISTORY_FILE = '~/.pywbemcli_history'

#: Default uri scheme if none is provided.  Thus if a server uri without
#: scheme component is provided, this is the default prepended to the
#: uri.

DEFAULT_URL_SCHEME = 'https'

#: Default pull MaxObjectCount if none is provided.  This is the maximum
#: number of objects per request that will be returned if the server
#: uses the pull operations for EnumerateInstances, AssociatorInstances,
#: etc. Set to the same default as used by pywbem.
DEFAULT_MAXPULLCNT = 1000

#: Maximum allowed connection timeout in seconds.  The environment will not
#: allow a connection timeout value larger than this on the command line or
#: internal option for timeout.
MAX_TIMEOUT = 300

#: If True, the table formatter uses the terminal width as the maximum width
#: of a table. If False it uses the DEFAULT_TABLE_WIDTH config variable. This
#: sets the maximum width of table output. The table formatter tries to
#: build any table output within this character width.
USE_TERMINAL_WIDTH = True

#: Default maximum character width of tables if USE_TERMINAL_WIDTH is NOT set.
#: If this variable is an integer, that is the maximum width. If None, tables
#: are output with no limit on width.
DEFAULT_TABLE_WIDTH = 150


#: If True, the auto-suggestion capability is enabled in the interactive
#: mode.  This capability uses the history file to provide suggestions for
#: the command file in addition to other auto-complete capabilities
#: If False, auto-suggestion is disabled.
USE_AUTOSUGGEST = True
